misc/custom_promise_types/cfengine.py

```diff
@@ -161,13 +208,78 @@ def _add_traceback_to_response(self):
         logs.append({"level": "debug", "message": trace})
         self._response["log"] = logs
 
+    def add_attribute(self, name, typing, default=None, required=False, default_to_promiser=False, validator=None):
+        attribute = OrderedDict()
+        attribute["name"] = name
+        attribute["typing"] = typing
+        attribute["default"] = default
+        attribute["required"] = required
+        attribute["default_to_promiser"] = default_to_promiser
+        attribute["validator"] = validator
+        self._validator_attributes[name] = attribute
+
+    @property
+    def _has_validation_attributes(self):
+        return bool(self._validator_attributes)
+
+    def create_attribute_object(self, promiser, attributes):
+
+        # Check for missing required attributes:
+        for name, attribute in self._validator_attributes.items():
+            if attribute["required"] and name not in attributes:
+                raise ValidationError(f"Missing required attribute '{a}'")
```

Comment:
I don't see an `a` variable defined here for use in the f-string.
Maybe you mean `{name}` instead?

Reply:
Indeed, thanks, I will fix in a follow-up.

Reply:
Addressed as first commit in: https://github.com/cfengine/core/pull/4700